### ISSUE TITLE
How to allow anonymous requests for introspection schema?

### ISSUE THREAD

**Issue by ropez** [graphql-aspnet/graphql-aspnet#156]
This might be a request for a pointer in the right direction, or it might be that we hit a limitation, I'm not sure.

Schenario: An ASP server setup that require authentication by default for all requests.

As I understand, it's possible to allow anonymous requests for specific GraphQL fields, by adding the `[AllowAnonymous]` attribute to a controller method.

How can we allow anonymous requests to the built-in `__schema` query? Since it's built-in, we don't have a controller where we can override the default settings for this query.

**Comment by kevin-carroll** [comment]
Hey @ropez !

Tricky scenario to be sure. I don't know if I can offer a solution but perhaps some insight.

TL;DR:  There isn't an easy solution to this. 

If you have independently set up your asp.net server to require authentication its likely that check is done in the http pipeline before the HttpContext is dispatched to a graphql middleware.  If this is the case the graphql runtime has no option to get a hold of the request and there won't be any way to allow the __schema query through unless you independently start checking the post body for the query. This can get pretty tricky if the supplied document has multiple queries and I don't recommend it.

Even once the graphql runtime gets ahold of the request there isn't any meaningful exposure of the schema object that you could modify.  Internally, `__schema` is just a field defined on the top level `query` object, there is nothing special about it and its treated with the same authority as any other fields, objects or controllers you add.  That is to say there is no hook or definition that says "if schema do X"; the default graphql pipeline literally doesn't know its executing a schema request or your custom code.  

All that said the pipelines are very configurable. As long as the request is making it to the graphql runtime you have the option to get in the middle of things and do what you need to do. You could rebuild the default graphql query pipeline and inject a component after the `AssignQueryOperationMiddleware` component at which point you could inspect the requested top level field set of the chosen operation and determine if the target is the schema and act accordingly.   This is a rather convoluted approach but could work and give you the most control and safety.  I'm happy to provide help on how to get started if you want to take this route, just let me know. 

Yet another option is to purposefully tag all your controllers with `[Authorize]` at the class level. The schema object allows anonymous access by default so as long as all your controllers require authorization and you aren't enforcing authentication at the asp.net level,  you should be able to setup the scenario you are looking for where all the "meaningful" data is gated but __schema is left open.  This has its own draw backs though since its a required opt in to security and not a an explicit opt out of schema.  You'd have to be diligent in the code you write to not expose yourself.

Hope this helps,
Kevin

P.S. I do think this is a valid use case and should be able to make this type of request a lot easier to handle in v2 of the library coming out later this year.

**Comment by kevin-carroll** [comment]
I should also add that if you are using the configuration option `QueryHandler.AuthenticatedRequestsOnly` thats essentially the same thing as an "asp.net level check". The GraphQL Htttp Middleware is registered with an [extra if block](https://github.com/graphql-aspnet/graphql-aspnet/blob/5cd17260dd1242845fd61bf798cb9f219c95f3e2/src/graphql-aspnet/Engine/SecureGraphQLHttpprocessor%7BTSchema%7D.cs#L52) to reject unauthenticated requests before they ever make it to the runtime for processing. Its doesn't take in to account the type of request before rejecting it.

**Comment by ropez** [comment]
Thank you Kevin for answering in so much detail.

I'm thinking that we can hopefully find a pragmatic solution to this. Perhaps we can set up the service that is doing the schema stitching with some kind of anonymous authentication token without access to any resources, to get around this.

PS: Thank you for creating this library. I think it generally works very well, and is a very good fit for our organization.
